src/compiler/crystal/macros/methods.cr

```diff
@@ -156,6 +158,27 @@ module Crystal
       end
     end
 
+    def interpret_parse_type(node)
+      interpret_check_args_toplevel do |arg|
+        arg.accept self
+        type_name = case last = @last
+                    when StringLiteral then last.value
+                    else
+                      arg.raise "argument to parse_type must be a StringLiteral, not #{last.class_desc}"
+                    end
+
+        arg.raise "argument to parse_type cannot be an empty value" if type_name.blank?
+
+        parser = Crystal::Parser.new type_name
+        parser.next_token
+        type = parser.parse_bare_proc_type
+        parser.check :EOF
+        @last = type
+      rescue ex : Crystal::SyntaxException
+        arg.raise "Invalid type name: #{type_name.inspect}"
```

Comment:
Another minor improvement: I think it makes sense to restrict the rescue to only the part where a syntax error can occur:
```suggestion
        begin
          parser = Crystal::Parser.new type_name
          parser.next_token
          type = parser.parse_bare_proc_type
          parser.check :EOF
          @last = type
        rescue ex : Crystal::SyntaxException
          arg.raise "Invalid type name: #{type_name.inspect}"
        end
```